File: Server/Modules/Utils/DAL/Common/Dtos/queryUtils.ts

export function renderQueryObject(queryParams: IQueryParams): IQueryObject {
    var result: IQueryObject = {};

    if (queryParams.keys != null) {
        result.keys = utils.GetKeyFromString(queryParams.keys);
    }

    if (queryParams.hasOwnProperty("select")) {
        result.select = queryParams.select.split(",").map((it: string) => it.trim());
    }

    if (queryParams.hasOwnProperty("filter")) {
        result.filter = queryParams.filter;
    }

    if (queryParams.hasOwnProperty("filterExpand")) {
        result.filterExpand = queryParams.filterExpand.split(";").map((it: string) => {
            var items = it.split(":").map((it) => it.trim());
            return {
                expand: items[0],
                filter: items[1] === "*" ? "1" : items[1]
            };
        });
    }

    if (queryParams.hasOwnProperty("orderBy")) {
        result.orderBy = queryParams.orderBy.split(",").map((it: string) => it.trim());
    }

    if (queryParams.hasOwnProperty("expand")) {
        result.expand = queryParams.expand.split(",").map((it: string) => it.trim());
    }

    if (queryParams.hasOwnProperty("count")) {
        result.count = Boolean(JSON.parse(queryParams.count));
    }

    if (queryParams.hasOwnProperty("skip") && !isNaN(parseInt(queryParams.skip))) {
        result.skip = parseInt(queryParams.skip);
    }

    if (queryParams.hasOwnProperty("top") && !isNaN(parseInt(queryParams.top))) {
        result.top = parseInt(queryParams.top);
    }

    return result;
}


export function renderQueryString(queryObject: IQueryObject): string {
    var result: string[] = [];

    if (!queryObject) {
        return "";
    }

    if (queryObject.hasOwnProperty("keys") && queryObject.keys.length) {
        result.push("keys=" + utils.GetStringFromKey(queryObject.keys));
    }

    if (queryObject.hasOwnProperty("select") && queryObject.select.length) {
        result.push("select=" + queryObject.select.join(","));
    }

    if (queryObject.hasOwnProperty("filter")) {
        result.push("filter=" + queryObject.filter);
    }

    if (queryObject.hasOwnProperty("filterExpand") && queryObject.filterExpand.length) {
        result.push("filterExpand=" + queryObject.filterExpand.map((it) => it.expand + ":" + it.filter || "*").join(","));
    }

    if (queryObject.hasOwnProperty("orderBy") && queryObject.orderBy.length) {
        result.push("orderBy=" + queryObject.orderBy.join(","));
    }

    if (queryObject.hasOwnProperty("expand") && queryObject.expand.length) {
        result.push("expand=" + queryObject.expand.join(","));
    }

    if (queryObject.hasOwnProperty("count")) {
        result.push("count=" + (queryObject.count ? "true" : "false"));
    }

    if (queryObject.hasOwnProperty("skip")) {
        result.push("skip=" + queryObject.skip.toString());
    }

    if (queryObject.hasOwnProperty("top")) {
        result.push("top=" + queryObject.top.toString());
    }

    return result.join("&");
}


module utils {

    /**
     * keys=key1:1,2,3,4;key2:4,5,6,7
     * ... will become:
     * keys=[ { key1: 1, key2: 4}, { key1: 2, key2: 5 }, { key1: 2, key2: 6 }, { key1: 4, key2: 7 } ]
     */
    export function GetKeyFromString(keys: string): IDictionary<any>[] {
        var keyValueSet: IDictionary<number[]> = {};
        var count = 0;
        for (let item of keys.split(";")) {
            var keyValue = item.split(":");
            keyValueSet[keyValue[0]] = keyValue[1].split(",").map((it) => parseInt(it));
            if (!count) {
                count = keyValueSet[keyValue[0]].length;
            } else {
                if (count != keyValueSet[keyValue[0]].length) {
                    throw "parametru 'keys' incorect";
                }
            }
        }

        var result: IDictionary<any>[] = [];
        for (let i = 0; i < count; i++) {
            var resultItem: IDictionary<number> = {};
            for (let keyName in keyValueSet) {
                resultItem[keyName] = keyValueSet[keyName][i];
            }
            result.push(resultItem);
        }
        return result;
    }

    /**
     * keys=[ { key1: 1, key2: 4}, { key1: 2, key2: 5 }, { key1: 2, key2: 6 }, { key1: 4, key2: 7 } ]
     * ... will become:
     * keys=key1:1,2,3,4;key2:4,5,6,7
     */
    export function GetStringFromKey(keys: IDictionary<any>[]): string {
        var keySet: IDictionary<any[]> = {};
        for (let dto of keys) {
            for (let keyName in dto) {
                if (!keySet.hasOwnProperty(keyName)) {
                    keySet[keyName] = [];
                }
                keySet[keyName].push(dto[keyName]);
            }
        }
        var result: string[] = [];
        for (let name in keySet) {
            result.push(`${name}:${keySet[name].join(",")}`);
        }
        return result.join(";");
    }

}

File: Server/Models/Utils/DAL/Common/Dtos/QueryUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace Server.Models.Utils.DAL.Common
{
    public static class QueryUtils
    {
        public static QueryObject RenderQueryObject(QueryParams queryParams)
        {
            var result = new QueryObject();

            if (queryParams.Keys != null)
            {
                result.Keys = GetKeyFromString(queryParams.Keys);
            }

            if (queryParams.Select != null)
            {
                result.Select = queryParams.Select.Split(new char[] { ',' }).Select((it) => it.Trim()).ToArray();
            }

            if (queryParams.Filter != null)
            {
                result.Filter = queryParams.Filter;
            }

            if (queryParams.FilterExpand != null)
            {
                result.FilterExpand = queryParams.FilterExpand.Split(new char[] { ';' }).Select((it) =>
                {
                    var items = it.Split(new char[] { ':' }).Select((it1) => it1.Trim()).ToArray();
                    return new FilterExpand()
                    {
                        Expand = items[0],
                        Filter = items[1] == "*" ? "1" : items[1]
                    };
                }).ToList();
            }

            if (queryParams.OrderBy != null)
            {
                result.OrderBy = queryParams.OrderBy.Split(new char[] { ',' }).Select((it) => it.Trim()).ToArray();
            }

            if (queryParams.Expand != null)
            {
                result.Expand = queryParams.Expand.Split(new char[] { ',' }).Select((it) => it.Trim()).ToArray();
            }

            if (queryParams.Count != null)
            {
                result.Count = Convert.ToBoolean(queryParams.Count);
            }

            if (queryParams.Skip != null && queryParams.Skip.IsNumeric())
            {
                result.Skip = Convert.ToInt32(queryParams.Skip);
            }

            if (queryParams.Top != null && queryParams.Top.IsNumeric())
            {
                result.Top = Convert.ToInt32(queryParams.Top);
            }

            return result;
        }

        public static string RenderQueryString(QueryObject queryObject)
        {
            var result = new List<string>();

            if (queryObject.Keys != null && queryObject.Keys.Length > 0)
            {
                result.Add("keys=" + GetStringFromKey(queryObject.Keys));
            }

            if (queryObject.Select != null && queryObject.Select.Length > 0)
            {
                result.Add("select=" + string.Join(",", queryObject.Select));
            }

            if (queryObject.Filter != null)
            {
                result.Add("filter=" + queryObject.Filter);
            }

            if (queryObject.FilterExpand != null && queryObject.FilterExpand.Count > 0)
            {
                result.Add("filterExpand=" + string.Join(",", queryObject.FilterExpand.Select((it) => it.Expand + ":" + (string.IsNullOrEmpty(it.Filter) ? "*" : it.Filter))));
            }

            if (queryObject.OrderBy != null && queryObject.OrderBy.Length > 0)
            {
                result.Add("orderBy=" + string.Join(",", queryObject.OrderBy));
            }

            if (queryObject.Expand != null && queryObject.Expand.Length > 0)
            {
                result.Add("expand=" + string.Join(",", queryObject.Expand));
            }

            if (queryObject.Count != null)
            {
                result.Add("count=" + ((bool)queryObject.Count ? "true" : "false"));
            }

            if (queryObject.Skip != null)
            {
                result.Add("skip=" + queryObject.Skip.ToString());
            }

            if (queryObject.Top != null)
            {
                result.Add("top=" + queryObject.Top.ToString());
            }

            return string.Join("&", result);
        }

        private static bool IsNumeric(this string input)
        {
            return Regex.IsMatch(input, @"^\d+$");
        }

        /**
         * keys=key1:1,2,3,4;key2:4,5,6,7
         * ... will become:
         * keys=[ { key1: 1, key2: 4}, { key1: 2, key2: 5 }, { key1: 2, key2: 6 }, { key1: 4, key2: 7 } ]
         */
        private static Dto[] GetKeyFromString(string keys)
        {
            var keyValueSet = new Dictionary<string, int[]>();
            var count = 0;
            foreach (var item in keys.Split(new char[] { ';' }))
            {
                var keyValue = item.Split(new char[] { ':' });
                keyValueSet.Add(keyValue[0], keyValue[1].Split(new char[] { ',' }).Select((it) => Convert.ToInt32(it)).ToArray());
                if (count == 0)
                {
                    count = keyValueSet[keyValue[0]].Length;
                }
                else
                {
                    if (count != keyValueSet[keyValue[0]].Length)
                    {
                        throw new ArgumentException("parametru 'keys' incorect");
                    }
                }
            }

            var result = new List<Dto>();
            for (var i = 0; i < count; i++)
            {
                var resultItem = new Dto();
                foreach (var keyValue in keyValueSet)
                {
                    resultItem.Add(keyValue.Key, keyValue.Value[i]);
                }
                result.Add(resultItem);
            }
            return result.ToArray();
        }

        /**
         * keys=[ { key1: 1, key2: 4}, { key1: 2, key2: 5 }, { key1: 2, key2: 6 }, { key1: 4, key2: 7 } ]
         * ... will become:
         * keys=key1:1,2,3,4;key2:4,5,6,7
         */
        private static string GetStringFromKey(Dto[] keys)
        {
            var keySet = new Dictionary<string, List<object>>();
            foreach (var dto in keys)
            {
                foreach (var item in dto)
                {
                    if (!keySet.ContainsKey(item.Key))
                    {
                        keySet.Add(item.Key, new List<object>());
                    }
                    keySet[item.Key].Add(item.Value);
                }
            }
            var result = new List<string>();
            foreach (var item in keySet)
            {
                result.Add(string.Format("{0}:{1}", item.Key, string.Join(",", item.Value)));
            }
            return string.Join(";", result);
        }
    }

}